qemu/qemu.go

```diff
@@ -26,14 +26,876 @@ package qemu
+func (object Object) Valid() bool {
+	switch object.Type {
+	case MemoryBackendFile:
+		if object.ID == "" || object.MemPath == "" || object.Size == 0 {
+			return false
+		}
+	}
+
+	return true
+}
+
+// FSDriver represents a qemu filesystem driver.
+type FSDriver string
+
+// SecurityModelType is a qemu filesystem security model type.
+type SecurityModelType string
+
+const (
+	// Local is the local qemu filesystem driver.
+	Local FSDriver = "local"
+
+	// Handle is the handle qemu filesystem driver.
+	Handle = "handle"
+
+	// Proxy is the proxy qemu filesystem driver.
+	Proxy = "proxy"
+)
+
+const (
+	// None is like passthrough without failure reports.
+	None SecurityModelType = "none"
+
+	// PassThrough uses the same credentials on both the host and guest.
+	PassThrough = "passthrough"
+
+	// MappedXattr stores some files attributes as extended attributes.
+	MappedXattr = "mapped-xattr"
+
+	// MappedFile stores some files attributes in the .virtfs directory.
+	MappedFile = "mapped-file"
+)
+
+// FSDevice represents a qemu filesystem configuration.
+type FSDevice struct {
+	// Driver is the filesystem driver backend.
+	Driver FSDriver
+
+	// ID is the filesystem identifier.
+	ID string
+
+	// Path is the host root path for this filesystem.
+	Path string
+
+	// MountTag is the device filesystem mount point tag.
+	MountTag string
+
+	// SecurityModel is the security model for this filesystem device.
+	SecurityModel SecurityModelType
+}
+
+// Valid returns true if the FSDevice structure is valid and complete.
+func (fsdev FSDevice) Valid() bool {
+	if fsdev.ID == "" || fsdev.Path == "" || fsdev.MountTag == "" {
+		return false
+	}
+
+	return true
+}
+
+// CharDeviceBackend is the character device backend for qemu
+type CharDeviceBackend string
+
+const (
+	// Pipe creates a 2 way connection to the guest.
+	Pipe CharDeviceBackend = "pipe"
+
+	// Socket creates a 2 way stream socket (TCP or Unix).
+	Socket = "socket"
+
+	// CharConsole sends traffic from the guest to QEMU's standard output.
+	CharConsole = "console"
+
+	// Serial sends traffic from the guest to a serial device on the host.
+	Serial = "serial"
+
+	// TTY is an alias for Serial.
+	TTY = "tty"
+
+	// PTY creates a new pseudo-terminal on the host and connect to it.
+	PTY = "pty"
+)
+
+// CharDevice represents a qemu character device.
+type CharDevice struct {
+	Backend CharDeviceBackend
+
+	ID   string
+	Path string
+}
+
+// Valid returns true if the CharDevice structure is valid and complete.
+func (cdev CharDevice) Valid() bool {
+	if cdev.ID == "" || cdev.Path == "" {
+		return false
+	}
+
+	return true
+}
+
+// NetDeviceType is a qemu networing device type.
+type NetDeviceType string
+
+const (
+	// TAP is a TAP networking device type.
+	TAP NetDeviceType = "tap"
+)
+
+// NetDevice represents a guest networking device
+type NetDevice struct {
+	// Type is the netdev type (e.g. tap).
+	Type NetDeviceType
+
+	// ID is the netdevice identifier.
+	ID string
+
```

Comment:
@mcastelino There already is a VHost boolean. Also, is the number of queues != len (NetDevice.FDs) ?